pennylane/ops/op_math/condition.py

```diff
@@ -430,3 +446,127 @@ def wrapper(*args, **kwargs):
         )
 
     return wrapper
+
+
+def _validate_abstract_values(
+    outvals: list, expected_outvals: list, branch_type: str, index: int = None
+) -> None:
+    """Ensure the collected abstract values match the expected ones."""
+
+    if len(outvals) != len(expected_outvals):
+        raise ValueError(
+            f"Mismatch in number of output variables in {branch_type} branch"
+            f"{'' if index is None else ' #' + str(index)}: "
+            f"{len(outvals)} vs {len(expected_outvals)}"
+        )
+
+    for i, (outval, expected_outval) in enumerate(zip(outvals, expected_outvals)):
+        if outval != expected_outval:
+            raise ValueError(
+                f"Mismatch in output abstract values in {branch_type} branch"
+                f"{'' if index is None else ' #' + str(index)} at position {i}: "
+                f"{outval} vs {expected_outval}"
+            )
+
+
+@functools.lru_cache
+def _get_cond_qfunc_prim():
+    """Get the cond primitive for quantum functions."""
+
+    import jax  # pylint: disable=import-outside-toplevel
+
+    cond_prim = jax.core.Primitive("cond")
+    cond_prim.multiple_results = True
+
+    @cond_prim.def_impl
+    def _(conditions, *args_and_consts, jaxpr_branches, n_consts_per_branch, n_args):
+
+        args = args_and_consts[:n_args]
+        consts_flat = args_and_consts[n_args:]
+
+        consts_per_branch = []
+        start = 0
+        for n in n_consts_per_branch:
+            consts_per_branch.append(consts_flat[start : start + n])
+            start += n
+
+        for pred, jaxpr, consts in zip(conditions, jaxpr_branches, consts_per_branch):
+            if pred and jaxpr is not None:
+                return jax.core.eval_jaxpr(jaxpr.jaxpr, consts, *args)

```

Comment:
If I'm reading this right, we should be able to combine these two loops?

Reply:
Yep! it's quite embarrassing that I missed this, but thanks for noticing